lib/location/method_gnss.c

```diff
@@ -866,6 +871,113 @@ static void method_gnss_print_pvt(const struct nrf_modem_gnss_pvt_data_frame *pv
 	}
 }
 
+#if defined(CONFIG_LOCATION_SERVICE_NRF_CLOUD_GNSS_POS_CLOUD_SEND)
+
+static int method_gnss_nrf_cloud_data_send(char *body)
+{
+	int ret;
+
+#if defined(CONFIG_NRF_CLOUD_MQTT)
+	struct nrf_cloud_tx_data mqtt_msg = {
+		.data.ptr = body,
+		.data.len = strlen(body),
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE,
+		.topic_type = NRF_CLOUD_TOPIC_MESSAGE,
+	};
+
+	ret = nrf_cloud_send(&mqtt_msg);
+	if (ret) {
+		LOG_ERR("MQTT: location data sending failed: %d", ret);
+	}
+#elif defined(CONFIG_NRF_CLOUD_COAP)
+	ret = nrf_cloud_coap_json_message_send(body, false, true);
+	if (ret) {
+		LOG_ERR("CoAP: location data sending failed");
+	}
+#elif defined(CONFIG_NRF_CLOUD_REST)
+#define REST_DETAILS_RX_BUF_SZ 300 /* No payload in a response, "just" headers */
+	static char rx_buf[REST_DETAILS_RX_BUF_SZ];
+	static char device_id[NRF_CLOUD_CLIENT_ID_MAX_LEN + 1];
+	static struct nrf_cloud_rest_context rest_ctx = {
+		.connect_socket = -1,
+		.keep_alive = false,
+		.rx_buf = rx_buf,
+		.rx_buf_len = sizeof(rx_buf),
+		.fragment_size = 0
+	};
+
+	ret = nrf_cloud_client_id_get(device_id, sizeof(device_id));
+	if (ret == 0) {
+		ret = nrf_cloud_rest_send_device_message(&rest_ctx, device_id, body, false, NULL);
+		if (ret) {
+			LOG_ERR("REST: location data sending failed: %d", ret);
+		}
+	} else {
+		LOG_ERR("Failed to get device ID, error: %d", ret);
+	}
+#endif
+	return ret;
+}
+
+static void method_gnss_nrf_cloud_pos_send(struct location_data *location_result)
+{
+	int err = 0;
+	char *json_str = NULL;
+	cJSON *gnss_data_obj = NULL;
+	struct nrf_cloud_gnss_data gnss_data = {
+		.type = NRF_CLOUD_GNSS_TYPE_PVT,
+		.pvt = {
+			.lon	  = location_result->longitude,
+			.lat	  = location_result->latitude,
+			.accuracy = location_result->accuracy,
```

Comment:
We could add more info here:
```suggestion
			.accuracy = location_result->accuracy,
			.speed =	pvt_data->speed,
			.has_speed =	location_result->flags & NRF_MODEM_GNSS_PVT_FLAG_VELOCITY_VALID,
			.heading =	location_result->heading,
			.has_heading =	location_result->flags & NRF_MODEM_GNSS_PVT_FLAG_VELOCITY_VALID,
```
Also, altitude might be available if we're not using low-accuracy mode,

Reply:
I implemented similar support for ATv2 some time ago. I didn't remember that there's a flag for velocity validity in the GNSS API. 😃 But, `NRF_MODEM_GNSS_PVT_FLAG_VELOCITY_VALID` does not tell whether heading is valid or not. For heading, I suggest something like this instead:

https://github.com/nrfconnect/sdk-nrf/blob/e5e367d13d164c3dcfb67e0bcac16111cdd5e4c0/applications/asset_tracker_v2/src/cloud/cloud_codec/nrf_cloud/nrf_cloud_codec.c#L167

Altitude is always valid and could be sent to the cloud. Low accuracy mode doesn't mean that altitude would not be valid. Valid altitude is needed to be able to calculate the position using only three satellites. If the (injected) altitude is incorrect, also the calculated position will have error. If we want, there could be some filtering for the altitude based on `altitude_accuracy`, but we don't do that in ATv2 currently.

Reply:
I'll need to pass PVT data into this function instead of location_result to get this done properly.

Reply:
This is now handled by using `nrf_modem_gnss_pvt_data_frame` and `NRF_CLOUD_GNSS_TYPE_MODEM_PVT`.

Heading issue has to be handled in `nrf_cloud_modem_pvt_data_encode()` if needed via separate PR.

Reply:
Because of CoAP CBOR, I need to go back to using `NRF_CLOUD_GNSS_TYPE_PVT`.